analyzers/src/SonarAnalyzer.Common/Rules/HardcodedIpAddressBase.cs

```diff
@@ -20,74 +20,130 @@
+                }
+
+                return supportedDiagnostics;
+            }
+        }
+
         protected HardcodedIpAddressBase(IAnalyzerConfiguration analyzerConfiguration) : base(analyzerConfiguration)
         {
         }
 
-        protected Action<SyntaxNodeAnalysisContext> GetAnalysisAction(DiagnosticDescriptor rule) =>
-            c =>
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(GeneratedCodeRecognizer, AnalyzeLiteral, SyntaxKind);
+
+        private void AnalyzeLiteral(SyntaxNodeAnalysisContext context)
+        {
+            if (!IsEnabled(context.Options))
             {
-                var stringLiteral = (TLiteralExpression)c.Node;
-                var literalValue = GetValueText(stringLiteral);
-                if (IsLocalHost(literalValue))
-                {
-                    return;
-                }
+                return;
+            }
 
-                var variableName = GetAssignedVariableName(stringLiteral);
-                if (variableName != null &&
-                    IgnoredVariableNames.Any(variableName.Contains))
-                {
-                    return;
-                }
+            CheckForHardcodedIpAddresses(context);
+        }
 
-                if (HasAttributes(stringLiteral))
-                {
-                    return;
-                }
+        private void CheckForHardcodedIpAddresses(SyntaxNodeAnalysisContext context)
+        {
+            var stringLiteral = (TLiteralExpression)context.Node;
+            var variableName = GetAssignedVariableName(stringLiteral);
 
-                if (!IPAddress.TryParse(literalValue, out var address))
-                {
-                    return;
-                }
+            if (variableName != null
+                && ignoredVariableNames.Any(variableName.Contains))
+            {
+                return;
+            }
 
-                if (address.AddressFamily == AddressFamily.InterNetwork &&
-                    literalValue.Split('.').Length != 4)
-                {
-                    return;
-                }
+            if (HasAttributes(stringLiteral))
+            {
+                return;
+            }
 
-                c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, stringLiteral.GetLocation(), literalValue));
-            };
+            var literalValue = GetValueText(stringLiteral);
+
+            if (!IPAddress.TryParse(literalValue, out var address))
+            {
+                return;
+            }
+
+            if (IsLoopbackAddress(literalValue) || IsNonRoutableAddress(literalValue) || IsBroadcast(literalValue) || DoesItLookLikeObjectIdentifier(literalValue))
+            {
+                return;
+            }
+
+            if (address.AddressFamily == AddressFamily.InterNetwork
+                && literalValue.Split('.').Length != NumOfIPv4AddressParts)
```

Comment:
This should be cheaper, since we won't generate the intermediate string array
```suggestion
                && literalValue.Count(x => x == '.') != NumOfIPv4AddressParts - 1)
```